### ISSUE TITLE
Code path in utils emits `jsonschema` deprecation warning

### ISSUE THREAD

**Issue by cdkini** [altair-viz/altair#3097]
[jsonschema](https://pypi.org/project/jsonschema/) recently released a new version and introduced a new deprecation warning around `jsonschema.RefResolver` - this is utilized in `altair/utils/schemapi.py`.

You can replicate with the following code:
```python
import altair as alt

alt.StandardType("...")
```

```
/opt/hostedtoolcache/Python/3.8.17/x64/lib/python3.8/site-packages/altair/vegalite/v4/schema/core.py:15771: in __init__
    super(StandardType, self).__init__(*args)
/opt/hostedtoolcache/Python/3.8.17/x64/lib/python3.8/site-packages/altair/utils/schemapi.py:197: in __init__
    self.to_dict(validate=True)
/opt/hostedtoolcache/Python/3.8.17/x64/lib/python3.8/site-packages/altair/utils/schemapi.py:358: in to_dict
    self.validate(result)
/opt/hostedtoolcache/Python/3.8.17/x64/lib/python3.8/site-packages/altair/utils/schemapi.py:462: in validate
    resolver = jsonschema.RefResolver.from_schema(cls._rootschema or cls._schema)
/opt/hostedtoolcache/Python/3.8.17/x64/lib/python3.8/site-packages/jsonschema/__init__.py:42: in __getattr__
    warnings.warn(
E   DeprecationWarning: jsonschema.RefResolver is deprecated as of v4.18.0, in favor of the https://github.com/python-jsonschema/referencing library, which provides more compliant referencing behavior as well as more flexible APIs for customization. A future release will remove RefResolver. Please file a feature request (on referencing) if you are missing an API for the kind of customization you need.
```

---

Please follow these steps to make it more efficient to solve your issue:

- [x] Since Altair is a Python wrapper around the Vega-Lite visualization grammar, [most bugs should be reported directly to Vega-Lite](https://github.com/vega/vega-lite/issues). You can click the Action Button of your Altair chart and "Open in Vega Editor" to create a reproducible Vega-Lite example and see if you get the same error in the Vega Editor.
- [x] Search for duplicate issues.
- [x] Use the latest version of Altair.
- [x] Describe how to reproduce the bug and include the full code and data to reproduce it, ideally using a sample data set from `vega_datasets`.

**Comment by binste** [comment]
Thank you @cdkini for letting us know! I'll look into this.

**Comment by binste** [comment]
I've been looking into how we can replace `RefResolver` with the new `referencing` library, spending quite some time reading the `referencing` and `jsonschema` docs (including [Migrating from RefResolver](https://python-jsonschema.readthedocs.io/en/stable/referencing/#migrating-from-refresolver)) but I'm somehow stuck. @Julian if you have a few minutes I'd highly appreciate your input. To make it easier, I slimmed it down to a minimal example.

# Setup
```python
import altair as alt
import jsonschema.validators
import referencing

# alt.load_schema returns the Vega-Lite schema as a dictionary.
# You can also view it at https://vega.github.io/schema/vega-lite/v5.json
rootschema = alt.load_schema()
schema = {'$ref': '#/definitions/PositionFieldDef'}
spec = {'field': 'x', 'type': 'temporal'}
```

# Our current approach (working)
```python
validator_cls = jsonschema.validators.validator_for(rootschema)
resolver = jsonschema.RefResolver.from_schema(rootschema)
validator = validator_cls(schema, resolver=resolver)

# This passes as expected
validator.validate(spec)

# This raises a ValidationError as expected: `Additional properties are not allowed ('test' was unexpected)`
validator.validate({**spec, "test": 3})
```

# Attempt to replace RefResolver with referencing library
```python
resource = referencing.Resource.from_contents(rootschema)
# The root schema does not have an ID and so I assumed that I can use "" but maybe this is
# where I'm wrong?
registry = referencing.Registry().with_resource(uri="", resource=resource)
validator_cls = jsonschema.validators.validator_for(rootschema)
validator = validator_cls(schema, registry=registry)

# This gives me the traceback below
validator.validate(spec)
```

```python
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File [~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/referencing/_core.py:186](https://file+.vscode-resource.vscode-cdn.net/Users/stefanbinder/Library/CloudStorage/Dropbox/Programming/altair/~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/referencing/_core.py:186), in Resource.pointer(self, pointer, resolver)
    185 try:
--> 186     contents = contents[segment]  # type: ignore[reportUnknownArgumentType]  # noqa: E501
    187 except LookupError:

KeyError: 'definitions'

During handling of the above exception, another exception occurred:

PointerToNowhere                          Traceback (most recent call last)
File [~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/jsonschema/validators.py:439](https://file+.vscode-resource.vscode-cdn.net/Users/stefanbinder/Library/CloudStorage/Dropbox/Programming/altair/~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/jsonschema/validators.py:439), in create..Validator._validate_reference(self, ref, instance)
    438 try:
--> 439     resolved = self._resolver.lookup(ref)
    440 except referencing.exceptions.Unresolvable as err:

File [~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/referencing/_core.py:594](https://file+.vscode-resource.vscode-cdn.net/Users/stefanbinder/Library/CloudStorage/Dropbox/Programming/altair/~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/referencing/_core.py:594), in Resolver.lookup(self, ref)
    593     resolver = self._evolve(registry=retrieved.registry, base_uri=uri)
--> 594     return retrieved.value.pointer(pointer=fragment, resolver=resolver)
    596 if fragment:

File [~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/referencing/_core.py:188](https://file+.vscode-resource.vscode-cdn.net/Users/stefanbinder/Library/CloudStorage/Dropbox/Programming/altair/~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/referencing/_core.py:188), in Resource.pointer(self, pointer, resolver)
    187 except LookupError:
--> 188     raise exceptions.PointerToNowhere(ref=pointer, resource=self)
    190 segments.append(segment)

PointerToNowhere: '/definitions/PositionFieldDef' does not exist within {'$ref': '#/definitions/PositionFieldDef'}

During handling of the above exception, another exception occurred:

_WrappedReferencingError                  Traceback (most recent call last)
Cell In[10], line 2
      1 # Works as expected
----> 2 validator.validate(spec)

File [~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/jsonschema/validators.py:427](https://file+.vscode-resource.vscode-cdn.net/Users/stefanbinder/Library/CloudStorage/Dropbox/Programming/altair/~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/jsonschema/validators.py:427), in create..Validator.validate(self, *args, **kwargs)
    426 def validate(self, *args, **kwargs):
--> 427     for error in self.iter_errors(*args, **kwargs):
    428         raise error

File [~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/jsonschema/validators.py:361](https://file+.vscode-resource.vscode-cdn.net/Users/stefanbinder/Library/CloudStorage/Dropbox/Programming/altair/~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/jsonschema/validators.py:361), in create..Validator.iter_errors(self, instance, _schema)
    358     continue
    360 errors = validator(self, v, instance, _schema) or ()
--> 361 for error in errors:
    362     # set details if not already set by the called fn
    363     error._set(
    364         validator=k,
    365         validator_value=v,
   (...)
    368         type_checker=self.TYPE_CHECKER,
    369     )
    370     if k not in {"if", "$ref"}:

File [~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/jsonschema/_validators.py:284](https://file+.vscode-resource.vscode-cdn.net/Users/stefanbinder/Library/CloudStorage/Dropbox/Programming/altair/~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/jsonschema/_validators.py:284), in ref(validator, ref, instance, schema)
    283 def ref(validator, ref, instance, schema):
--> 284     yield from validator._validate_reference(ref=ref, instance=instance)

File [~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/jsonschema/validators.py:441](https://file+.vscode-resource.vscode-cdn.net/Users/stefanbinder/Library/CloudStorage/Dropbox/Programming/altair/~/Library/CloudStorage/Dropbox/Programming/altair/.venv/lib/python3.10/site-packages/jsonschema/validators.py:441), in create..Validator._validate_reference(self, ref, instance)
    439         resolved = self._resolver.lookup(ref)
    440     except referencing.exceptions.Unresolvable as err:
--> 441         raise exceptions._WrappedReferencingError(err)
    443     return self.descend(
    444         instance,
    445         resolved.contents,
    446         resolver=resolved.resolver,
    447     )
    448 else:

_WrappedReferencingError: PointerToNowhere: '/definitions/PositionFieldDef' does not exist within {'$ref': '#/definitions/PositionFieldDef'}
```

# Where Altair currently uses RefResolver
Just for reference:
* https://github.com/altair-viz/altair/blob/b7ae4867a5c015a98a30cbe948d7ed35dad58ad1/altair/utils/schemapi.py#L118
* https://github.com/altair-viz/altair/blob/b7ae4867a5c015a98a30cbe948d7ed35dad58ad1/altair/utils/schemapi.py#L353
* https://github.com/altair-viz/altair/blob/b7ae4867a5c015a98a30cbe948d7ed35dad58ad1/tools/schemapi/utils.py#L16

**Comment by Julian** [comment]
Happy to have a look - will probably be Monday!

**Comment by Julian** [comment]
TL;DR: Yes you should assign some "real" URI to the Vega Lite schema I think (probably the one that already is the URL where it lives), and then use that URI to refer to your subschemas. Using the existing pointers in the Vega schema might be enough there, I'll try to show an example below.

Slightly less short:

When you say "`#/foo/bar`" you are saying "in the current schema resource is a key called foo with a key called bar inside it" -- so if you have a schema saying `{"$ref": "#/foo/bar"}` that's saying "in this schema is a `foo` key" as above -- but of course that isn't true -- there's just one key in that schema. The fact that what you had previously worked was sort of an "abuse" of the empty string doing some special stuff inside `RefResolver` which happened to use `""` as some internal fallback URI if it couldn't come up with one inside your root schema (which was outside the spec and wouldn't work if you tried it in some other JSON Schema implementation) -- so the error is trying to tell you exactly the above (there's no `definitions` key in the schema you're calling `""`.)

More specifically, some code that should do what your original code does:

```python
import warnings
with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    import altair as alt

import jsonschema.validators
import referencing

rooturi = "https://vega.github.io/schema/vega-lite/v5.json"
rootschema = alt.load_schema()
schema = {'$ref': f"{rooturi}#/definitions/PositionFieldDef"}
spec = {'field': 'x', 'type': 'temporal'}

# It looks like you were previously picking a subschema of the "real" schema
# that lives at this URL -- I'll point out you don't really need to do that at
# this point, you should be able to take the "real" root schema as-is and use
# fragments to reference inside of it. But I'm keeping things as-is in this
# snippet, other than adding a URI -- you can of course also pick another URI,
# like `urn:example:vega-lite-altair-schema` to refer to things!
registry = referencing.Registry().with_contents([(rooturi, rootschema)])
validator_cls = jsonschema.validators.validator_for(rootschema)
validator = validator_cls(schema, registry=registry)

validator.validate(spec)
validator.validate({**spec, "test": 3})  # gives your traceback as expected
``` 

Let me know if the above helps, and/or if you have suggestions. I can't commit fully to sending a PR making the changes to altair but I highly value the project so if there's help I can offer I can certainly say I'd like to help in any way I can, whether that's just a review of a PR, more comments, or sending a partial PR.
